server/src/main/java/org/elasticsearch/action/admin/cluster/state/TransportClusterStateAction.java

```diff
@@ -182,9 +194,56 @@ private void buildResponse(final ClusterStateRequest request,
                 }
             }
         }
-        listener.onResponse(new ClusterStateResponse(currentState.getClusterName(), builder.build(),
-            PublicationTransportHandler.serializeFullClusterState(currentState, Version.CURRENT).length(), false));
+
+        clusterStateSizeByVersionCache.getOrComputeCachedSize(currentState.version(),
+            () -> PublicationTransportHandler.serializeFullClusterState(currentState, Version.CURRENT).length(),
+            ActionListener.wrap(size ->
+                listener.onResponse(new ClusterStateResponse(currentState.getClusterName(), builder.build(), size, false)),
+                listener::onFailure));
     }
 
+    static class ClusterStateSizeByVersionCache {
+
+        // allow space for the sizes of the current and previous two cluster states to allow for out-of-order calls
+        static final int CACHE_SIZE = 3;
+
+        private final ThreadPool threadPool;
+
+        ClusterStateSizeByVersionCache(ThreadPool threadPool) {
+            this.threadPool = threadPool;
+        }
+
+        // The cluster state version might not completely determine its compressed size, because some changes (e.g. applying the no-master
+        // block) are made without updating its version. In practice it's close enough: most changes do update its version, and the
+        // size computed here is only used for reporting, so this is a reasonable approximation.
+        private Map<Long, PlainListenableActionFuture<Integer>> clusterStateSizeByVersionCache
+            = new LinkedHashMap<Long, PlainListenableActionFuture<Integer>>(CACHE_SIZE) {
+            @Override
+            protected boolean removeEldestEntry(Map.Entry eldest) {
+                return size() > CACHE_SIZE;
+            }
+        };
+
+        synchronized void getOrComputeCachedSize(final long clusterStateVersion,
```

Comment:
Maybe block a little less here, by simply using a `Collections.synchronizedMap()` on the map (that'll be safe with `computeIfAbsent`) since the `addListener` call is thread-safe anyway and it would probably be nice to have it resolve outside a `synchronized` block which currently it won't for cached values?

Reply:
`Collections.synchronizedMap()` doesn't let me bound the size of the map by overriding `removeEldestEntry()`. But you're right about calling `addListener` outside the mutex, I'll do that.

Reply:
@DaveCTurner you can just create the same map you're already creating and wrap it with `java.util.Collections#synchronizedMap`? :) That said it doesn't really matter, it's just less code than having your own mutex handling.

Reply:
Oh right I was thinking of `newConcurrentMap()`. Got it.